### PR TITLE
A2C example

### PR DESCRIPTION
**Pull request by zenghsh3**
- A3C algorithm
- A2C example

### CODE REVIEW

File: examples/A2C/README.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
+## Reproduce A2C with PARL
+Based on PARL, the A2C algorithm of deep reinforcement learning is reproduced, and the same level of indicators of the paper is reproduced in the classic Atari game.
+
+A2C is a synchronous, deterministic variant of [Asynchronous Advantage Actor Critic (A3C)](https://arxiv.org/abs/1602.01783)
+
+### Atari games introduction
+Please see [here](https://gym.openai.com/envs/#atari) to know more about Atari game.
+
+### Benchmark result
+Results with one learner (in P40 GPU) and 5 actors in 10 million sample steps.
```
Comment (TomorrowIsAnOtherDay) [inline examples/A2C/README.md]: `in 10 million sample steps.` should be corrected as `in around 2 hours`, because  x-axis in figures here represent `elapsed time`.

Comment (TomorrowIsAnOtherDay) [inline examples/A2C/README.md]: in a P40 GPU

File: examples/A2C/README.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    python actor.py &
+done;
+wait
+```
+
+You can change training settings (e.g. `env_name`, `server_ip`) in `a2c_config.py`.
+Training result will be saved in `log_dir/train/result.csv`.
+
+### Reference
++ [Ray](https://github.com/ray-project/ray)
```
Comment (TomorrowIsAnOtherDay) [inline examples/A2C/README.md:42]: Add a link for the openai's blog that introduced A2C?

File: examples/A2C/README.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
+## Reproduce A2C with PARL
+Based on PARL, the A2C algorithm of deep reinforcement learning is reproduced, and the same level of indicators of the paper is reproduced in the classic Atari game.
+
+A2C is a synchronous, deterministic variant of [Asynchronous Advantage Actor Critic (A3C)](https://arxiv.org/abs/1602.01783)
```
Comment (TomorrowIsAnOtherDay) [inline examples/A2C/README.md]: Add some notes here to explain why you import A3C from the parl.algorithms instead from A2C.

File: examples/A2C/README.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
+## Reproduce A2C with PARL
+Based on PARL, the A2C algorithm of deep reinforcement learning is reproduced, and the same level of indicators of the paper is reproduced in the classic Atari game.
```
Comment (TomorrowIsAnOtherDay) [inline examples/A2C/README.md]: Based on PARL, the A2C algorithm of deep reinforcement learning has been reproduced, reaching the same level of indicators as the paper in Atari benchmarks.

File: examples/A2C/actor.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+            for env_id in six.moves.range(self.config['env_num']):
+                env_sample_data[env_id]['obs'].append(self.obs_batch[env_id])
+                env_sample_data[env_id]['actions'].append(
+                    actions_batch[env_id])
+                env_sample_data[env_id]['rewards'].append(reward_batch[env_id])
+                env_sample_data[env_id]['dones'].append(done_batch[env_id])
+                env_sample_data[env_id]['values'].append(values_batch[env_id])
+
+                # Calculate advantages when episode done or reach max sample steps
```
Comment (TomorrowIsAnOtherDay) [inline examples/A2C/actor.py]: Calculate advantages when `the` episode `is` done or reach max sample steps

File: examples/A2C/learner.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+            if cnt % self.config['get_remote_metrics_interval'] == 0:
+                metrics = remote_actor.get_metrics()
+                if metrics:
+                    self.remote_metrics_queue.put(metrics)
+
+    def step(self):
+        """
+        1. kick off all actors to synchronize parameters and sample data;
+        2. collect sample data of all actors;
+        3. synchronize train and update parameters.
```
Comment (TomorrowIsAnOtherDay) [inline examples/A2C/learner.py]: synchronize train and update parameters. ?
It seems something wrong in this sentence and hard to understand it.
Please correct this line.

File: examples/A2C/atari_agent.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+        use_cuda = True if self.gpu_id >= 0 else False
+
+        exec_strategy = fluid.ExecutionStrategy()
+        exec_strategy.use_experimental_executor = True
+        exec_strategy.num_threads = 4
+        build_strategy = fluid.BuildStrategy()
+        build_strategy.remove_unnecessary_lock = True
+
+        # Use ParallelExecutor to make learn program running faster
```
Comment (TomorrowIsAnOtherDay) [inline examples/A2C/atari_agent.py]: Use ParallelExecutor to make learn program run faster.

File: examples/PPO/utils.py [hunk 8]
```diff
@@ -6,22 +6,7 @@
 import numpy as np
 import scipy.signal
```
Comment (TomorrowIsAnOtherDay) [inline examples/PPO/utils.py:8]: rename this file as `scaler.py`.

File: examples/A2C/atari_model.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+        flatten = layers.flatten(conv3, axis=1)
+        values = self.value_fc(flatten)
+        values = layers.squeeze(values, axes=[1])
+        return values
+
+    def policy_and_value(self, obs):
+        """
+        Args:
+            obs: An float32 tensor of shape [B, C, H, W]
```
Comment (TomorrowIsAnOtherDay) [inline examples/A2C/atari_model.py]: A instead An.

File: examples/A2C/README.md (hunk 1 above)
Comment (zenghsh3) [reply examples/A2C/README.md]: ”10 million sample steps” mentioned here is to provide information about data efficiency.

Comment (zenghsh3) [reply examples/A2C/README.md]: done

File: examples/A2C/README.md (hunk 2 above)
Comment (zenghsh3) [reply examples/A2C/README.md:42]: done

File: examples/A2C/README.md (hunk 3 above)
Comment (zenghsh3) [reply examples/A2C/README.md]: done

File: examples/A2C/README.md (hunk 4 above)
Comment (zenghsh3) [reply examples/A2C/README.md]: done

File: examples/A2C/actor.py (hunk 5 above)
Comment (zenghsh3) [reply examples/A2C/actor.py]: done

File: examples/A2C/learner.py (hunk 6 above)
Comment (zenghsh3) [reply examples/A2C/learner.py]: change to "update parameters"

File: examples/A2C/atari_agent.py (hunk 7 above)
Comment (zenghsh3) [reply examples/A2C/atari_agent.py]: done

File: examples/A2C/atari_model.py (hunk 9 above)
Comment (zenghsh3) [reply examples/A2C/atari_model.py]: done

File: examples/PPO/utils.py (hunk 8 above)
Comment (zenghsh3) [reply examples/PPO/utils.py:8]: done

**Review by TomorrowIsAnOtherDay — APPROVED** [review]
Great Job!
